cmd/podman/start.go

```diff
@@ -104,44 +96,41 @@ func startCmd(c *cli.Context) error {
 		if err != nil {
 			return errors.Wrapf(err, "unable to parse annotations in %s", ctr.ID())
 		}
+
+		// Handle start --attach
 		// We only get a terminal session if both a tty was specified in the spec and
 		// -a on the command-line was given.
 		if attach && tty {
-			// We increment the wg counter because we need to do the attach
-			wg.Add(1)
-			// Attach to the running container
-			go func() {
-				logrus.Debugf("trying to attach to the container %s", ctr.ID())
-				defer wg.Done()
-				if err := ctr.Attach(noStdIn, c.String("detach-keys"), attached); err != nil {
-					logrus.Errorf("unable to attach to container %s: %q", ctr.ID(), err)
-				}
-			}()
-			if !<-attached {
-				return errors.Errorf("unable to attach to container %s", ctr.ID())
+			attachChan, err := ctr.StartAndAttach(noStdIn, c.String("detach-keys"))
+			if err != nil {
+				return errors.Wrapf(err, "unable to start container %s", ctr.ID())
+			}
+
+			// Wait for attach to complete
+			err = <-attachChan
+			if err != nil {
+				return errors.Wrapf(err, "error attaching to container %s", ctr.ID())
 			}
+
+			if ecode, err := ctr.ExitCode(); err != nil {
+				logrus.Errorf("unable to get exit code of container %s: %q", ctr.ID(), err)
+			} else {
+				exitCode = int(ecode)
+			}
+
+			return ctr.Cleanup()
 		}
-		err = ctr.Start()
-		if err != nil {
+
+		// Handle non-attach start
+		if err := ctr.Start(); err != nil {
 			if lastError != nil {
 				fmt.Fprintln(os.Stderr, lastError)
 			}
-			lastError = errors.Wrapf(err, "unable to start %s", container)
+			lastError = errors.Wrapf(err, "unable to start container %q", container)
 			continue
 		}
-		if !attach {
-			fmt.Println(ctr.ID())
-		}
-		wg.Wait()
-		if ecode, err := ctr.ExitCode(); err != nil {
-			logrus.Errorf("unable to get exit code of container %s: %q", ctr.ID(), err)
-		} else {
-			exitCode = int(ecode)
-		}
-		if lastError != nil {
-			fmt.Fprintln(os.Stderr, lastError)
-		}
-		lastError = ctr.Cleanup()
+		fmt.Println(ctr.ID())
```

Comment:
Your always printing the ctr.ID?  even when you attach?

Reply:
agree that should only be when -d

Reply:
All the start --attach code is in the `if attach & tty {` block above, which always exits - this will only execute in non-attach cases

Reply:
Ok I messed it. LGTM